test/opengl/testing.h

```diff
@@ -0,0 +1,93 @@
+#ifndef _TESTING_H_
+#define _TESTING_H_
+
+#include "Halide.h"
+#include <iostream>
+#include <exception>
+#include <functional>
+#include <cmath>
+
+namespace Testing {
+
+    // fill 3-dimension buffer
+    template <typename T>
+        void fill(Halide::Buffer<T> &buf, std::function<T(int x, int y, int c)> f)
+        {
+            buf.for_each_element([&](int x, int y, int c) { 
+                    buf(x, y, c) = f(x, y, c);
+                    });
+        }
+
+    template <typename T>
+        bool neq(T a, T b, T tol)
+        {
+            return std::abs(a-b) > tol;
+        }
+
+    // check 3-dimension buffer
+    template<typename T>
+        bool check_result(const Halide::Buffer<T> &buf, std::function<T(int x, int y, int c)> f, T tol=0)
+        {
+            class err : std::exception {
+                public: static void vector(const std::vector<T> &v) {
+                    for (size_t i=0; i<v.size(); i++) {
+                        if (i > 0) std::cerr << ",";
+                        std::cerr << v[i]+0; // need to add 0 to get output -- compiler bug??
```

Comment:
Ah yeah.  Fails when T is `uint8_t` since that's not actually an integer it's a char, and so gets output as a (generally garbage) ascii character.   Simpler fix then is to use `+v[i]` so that the unary + operator will do the promotion to integer.   Switched to that and updated the comment.